### PR TITLE
Closes #4299 Add Mechanism to Keep Higher Priority Breadcrumbs if Max…

### LINKED ISSUE #4299: lib-crash BreadCrumbs Should Be Dropped Using Priority
Current lib-crash breadcrumb implementation have a maximum of 20 breadcrumbs stored.  The breadcrumbs are dropped in the FIFO order.  This implementation could be changed to use the breadcrumb priority value so priority + age is used to decide which breadcrumb is first dropped to make room for the new breadcrumb.

### PR DESCRIPTION
**Pull request by rocketsroger**
… Number of Breadcrumb is recorded

---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt [hunk 1]
```diff
@@ -97,4 +100,35 @@ data class Breadcrumb(
...
+        return result
+    }
+
+    fun toSortedArrayList(): ArrayList<Breadcrumb> {
+        val breadcrumbsArrayList: ArrayList<Breadcrumb> = arrayListOf()
+        if (isNotEmpty()) {
+            breadcrumbsArrayList.addAll(this)
+
+            /* Sort by timestamp before reporting */
+            breadcrumbsArrayList.sortWith(Comparator { bc0, bc1 -> bc0.date.compareTo(bc1.date) })
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt]: nit: Another option would be to use `sortBy {}` and "map" to a value for sorting (e.g. the date object or the timestamp).

File: components/lib/crash/src/test/java/mozilla/components/lib/crash/BreadcrumbTest.kt [hunk 2]
```diff
@@ -163,13 +163,121 @@ class BreadcrumbTest {
...
+        val testCategory = "testing_category"
+        val testType = Breadcrumb.Type.USER
+
+        val service = object : CrashReporterService {
+            override fun report(crash: Crash.UncaughtExceptionCrash) {
+            }
+
+            override fun report(crash: Crash.NativeCodeCrash) {
+            }
+        }
```
Comment (pocmo) [inline components/lib/crash/src/test/java/mozilla/components/lib/crash/BreadcrumbTest.kt]: nit: If you only need an empty implementation then a mock may be sufficient?

```kotlin
al service : CrashReporterService = mock()
```

File: components/lib/crash/src/test/java/mozilla/components/lib/crash/handler/ExceptionHandlerTest.kt [hunk 3]
```diff
@@ -62,9 +62,20 @@ class ExceptionHandlerTest {
...
 
+        val crashReporter = CrashReporter(
+                shouldPrompt = CrashReporter.Prompt.NEVER,
+                services = listOf(object : CrashReporterService {
+                    override fun report(crash: Crash.UncaughtExceptionCrash) {
+                    }
+
+                    override fun report(crash: Crash.NativeCodeCrash) {
+                    }
+                })
```
Comment (pocmo) [inline components/lib/crash/src/test/java/mozilla/components/lib/crash/handler/ExceptionHandlerTest.kt]: nit:` listOf(mock())`

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt [hunk 4]
```diff
@@ -97,4 +100,35 @@ data class Breadcrumb(
...
+    override fun compareTo(other: Breadcrumb): Int {
+        if (this.level.ordinal == other.level.ordinal) {
+            return this.date.compareTo(other.date)
+        }
+
+        return this.level.ordinal.compareTo(other.level.ordinal)
+    }
+}
+
+internal class BreadCrumbPriorityQueue(
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt]: I was torn between creating one or not.  Thanks for helping me make the decision. 👍   I'll move it into its own file.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt [hunk 5]
```diff
@@ -97,4 +100,35 @@ data class Breadcrumb(
...
+    private val maxSize: Int
+) : PriorityQueue<Breadcrumb>() {
+    override fun add(element: Breadcrumb?): Boolean {
+        val result = super.add(element)
+        if (this.size > maxSize) {
+            this.poll()
+        }
+        return result
+    }
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt]: You're probably right.  I'll make it synchronized.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt [hunk 6]
```diff
@@ -97,4 +100,35 @@ data class Breadcrumb(
...
+
+    fun toSortedArrayList(): ArrayList<Breadcrumb> {
+        val breadcrumbsArrayList: ArrayList<Breadcrumb> = arrayListOf()
+        if (isNotEmpty()) {
+            breadcrumbsArrayList.addAll(this)
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt]: Agreed.  Should be synchronized.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt (hunk 1 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt]: I'll look into if sortBy is a better solution.  Thanks,

File: components/lib/crash/src/test/java/mozilla/components/lib/crash/BreadcrumbTest.kt (hunk 2 above)
Comment (rocketsroger) [reply components/lib/crash/src/test/java/mozilla/components/lib/crash/BreadcrumbTest.kt]: I'll try it.

File: components/lib/crash/src/test/java/mozilla/components/lib/crash/handler/ExceptionHandlerTest.kt (hunk 3 above)
Comment (rocketsroger) [reply components/lib/crash/src/test/java/mozilla/components/lib/crash/handler/ExceptionHandlerTest.kt]: Sure I'll try it.

File: components/lib/crash/src/test/java/mozilla/components/lib/crash/BreadcrumbTest.kt (hunk 2 above)
Comment (rocketsroger) [reply components/lib/crash/src/test/java/mozilla/components/lib/crash/BreadcrumbTest.kt]: Works great thanks!

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt (hunk 1 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/Breadcrumb.kt]: Yes, you're right.  sortBy is so much cleaner.  Tested again and functionality did not change.  Thanks,

**Comment by rocketsroger** [comment]
bors r=pocmo

**Comment by rocketsroger** [comment]
bors retry

**Pull request merged**